lib/ClangImporter/DWARFImporter.cpp

```diff
@@ -0,0 +1,178 @@
+
+  virtual void
+  lookupClassMembers(ModuleDecl::AccessPathTy accessPath,
+                     VisibleDeclConsumer &consumer) const override {}
+
+  virtual void
+  lookupClassMember(ModuleDecl::AccessPathTy accessPath, DeclName name,
+                    SmallVectorImpl<ValueDecl *> &decls) const override {}
+
+  void lookupObjCMethods(
+      ObjCSelector selector,
+      SmallVectorImpl<AbstractFunctionDecl *> &results) const override {}
+
+  virtual void
+  getTopLevelDecls(SmallVectorImpl<Decl *> &results) const override {}
+
+  virtual void
+  getDisplayDecls(SmallVectorImpl<Decl *> &results) const override {}
+
+  virtual void
+  getImportedModules(SmallVectorImpl<ModuleDecl::ImportedModule> &imports,
+                     ModuleDecl::ImportFilter filter) const override {}
+
+  virtual void getImportedModulesForLookup(
+      SmallVectorImpl<ModuleDecl::ImportedModule> &imports) const override {};
+
+  virtual void collectLinkLibraries(
+      ModuleDecl::LinkLibraryCallback callback) const override {};
+
+  Identifier
+  getDiscriminatorForPrivateValue(const ValueDecl *D) const override {
+    llvm_unreachable("no private decls in Clang modules");
+  }
+
+  virtual StringRef getFilename() const override { return ""; }
+
+  virtual const clang::Module *getUnderlyingClangModule() const override {
+    return nullptr;
+  }
+
+  static bool classof(const FileUnit *file) {
+    return file->getKind() == FileUnitKind::DWARFModule;
+  }
+  static bool classof(const DeclContext *DC) {
+    return isa<FileUnit>(DC) && classof(cast<FileUnit>(DC));
+  }
+};
+
+ModuleDecl *ClangImporter::Implementation::loadModuleDWARF(
+    SourceLoc importLoc, ArrayRef<std::pair<Identifier, SourceLoc>> path) {
+  // There's no importing from debug info if no importer is installed.
+  if (!DWARFImporter)
+    return nullptr;
+
+  // FIXME: Implement submodule support!
+  Identifier name = path[0].first;
+  auto it = DWARFModuleUnits.find(name);
+  if (it != DWARFModuleUnits.end())
+    return it->second->getParentModule();
+
+  auto *decl = ModuleDecl::create(name, SwiftContext);
+  decl->setIsNonSwiftModule();
+  decl->setHasResolvedImports();
+  auto wrapperUnit = new (SwiftContext) DWARFModuleUnit(*decl, *this);
+  DWARFModuleUnits.insert({name, wrapperUnit});
+  decl->addFile(*wrapperUnit);
+
+  // Force load overlay modules for all imported modules.
+  decl->forAllVisibleModules({}, [](ModuleDecl::ImportedModule import) {});
+
+  // Register the module with the ASTContext so it is available for lookups.
+  ModuleDecl *&loaded = SwiftContext.LoadedModules[name];
+  if (!loaded)
+    loaded = decl;
+
+  return decl;
+}
+
+void ClangImporter::Implementation::lookupValueDWARF(
+    ModuleDecl::AccessPathTy accessPath, DeclName name, NLKind lookupKind,
+    SmallVectorImpl<ValueDecl *> &results) {
+  if (!swift::ModuleDecl::matchesAccessPath(accessPath, name))
+    return;
+
+  if (lookupKind != NLKind::QualifiedLookup)
```

Comment:
This seems wrong.

Reply:
This is something I meant to ask: What does an unqualified lookup mean in this function?Based on the name I would expect an unqualified lookup to also take some form of context that the unqualified name is relative to, but I don't see that (or is that the access path?). Does this mean something different?

Reply:
This is about how the lookup was originally written: was it `Module.Foo`, or just `Foo` and we haven't found anything in the current module? The only real reason we have it is so that the magic Builtin module can refuse to provide results for unqualified lookup.